### DISCUSSION TITLE
Question about Data object creation

### DISCUSSION THREAD

**Discussion by aalaily** [spatie/laravel-data#762] [Q&A]
Hello,

I have a question about how Data objects get created when referenced in a route parameter.

Based on this very simple code:
```
class UserData extends Data {
    public function __construct( public int $id, public string $email) {}
}
```
```
class User extends Model { protected $fillable = []; protected $guarded = []; }
```
```
class UserController extends Controller {
    public function getUser(UserData $user) {
        dd($user->toArray());
    }
}
```
```
Route::get('/admin/{user}', [UserController::class, 'getUser'])->name('user.get');
```

When someone visits ```http://localhost/admin/getUser/1``` should I expect that:

1.  Laravel loads User model from id 1
2. ```$user``` data object gets created from loaded User model
3. ```$user``` gets validated
4. The Controller function executes with a valid ```$user``` data object

OR

1. Laravel instantiates data object with id 1, as ```$user```
2. ```$user``` validates but validation fails because the required filed ```email``` is missing (only the id is passed in the route parameter)
3. Data object ```$user``` never gets created
4. The Controller function never executes and the application returns a validation error

Right now, it looks like the first scenario (where I get a valid ```$user``` data object in my Controller) is only possible if I change the signature of the controller function to ```public function getUser(User $user) {...}``` and then manually get the ```$userData``` data object from inside the function using something like ```UserData::from($user)```.

Thanks for your insight.

**Comment by rubenvanassche** [comment]
None of both of the cases, since we only create data from request values. Since none are provided the injection isn't working. You could try using the route parameter injection but the solution you're using now is in my opinion the way to go.

**Comment by CoyCoding** [comment]
If you are running into the issue where it fails validation but doesn't return the normal validation response. 

In the global error handler `Handler.php` you can register a catch all for `ValidationException` 

```
    public function register()
    {
        $this->renderable(function (ValidationException $e, Request $request) {
            return response()->json([
                'message' => $e->getMessage(),
                'errors' => $e->errors(),
            ], 422);
        });
    }
```

Unsure if this is the exact issue you were running into that made you question this, but for my API only instance of laravel it solved my problem.
